src/sulguk2/mapper.py

```diff
@@ -0,0 +1,298 @@
+import urllib.parse
+from functools import cached_property, partial
+from typing import Any, Callable, List, Optional, Tuple
+
+from sulguk.entities import (
+    Blockquote,
+    Bold,
+    Code,
+    Emoji,
+    Entity,
+    Group,
+    HorizontalLine,
+    Italic,
+    Link,
+    ListGroup,
+    ListItem,
+    NewLine,
+    Paragraph,
+    Pre,
+    Progress,
+    Quote,
+    Spoiler,
+    Strikethrough,
+    Stub,
+    Text,
+    Underline,
+    Uppercase,
+    ZeroWidthSpace,
+)
+from sulguk.render.numbers import NumberFormat
+
+Attrs = List[Tuple[str, str]]  # html5lib returns empty string if no value
+
+OL_FORMAT = {
+    "1": NumberFormat.DECIMAL,
+    "a": NumberFormat.LETTERS_LOWER,
+    "A": NumberFormat.LETTERS_UPPER,
+    "i": NumberFormat.ROMAN_LOWER,
+    "I": NumberFormat.ROMAN_UPPER,
+}
+
+LANG_CLASS_PREFIX = "language-"
+
+NO_CLOSING_TAGS = ("br", "wbr", "hr", "meta", "link", "img", "input")
+
+
+class Mapper:
+    def __init__(self, base_url: str | None = None) -> None:
+        self._base_url = base_url
+
+    def match(
+        self,
+        tag: str,
+        attrs: Attrs,
+    ) -> Tuple[Optional[Entity], Optional[Entity]]:
+        factory = self._map.get(tag)
+        if factory is None:
+            raise ValueError(f"Unsupported tag: {tag}")
+        result = factory(attrs)
+        if isinstance(result, tuple):
+            inner_entity, outer_entity = result
+        else:
+            inner_entity = None
+            outer_entity = result
+
+        return inner_entity, outer_entity
+
+    @cached_property
+    def _map(
+        self,
+    ) -> dict[str, Callable[[Attrs], Tuple[Entity, Entity] | Entity | None]]:
+        def _no_attrs(factory):
+            return lambda attrs: factory()
+
+        _map = {
+            # single tags
+            "br": _no_attrs(NewLine),
+            "wbr": _no_attrs(ZeroWidthSpace),
+            "hr": _no_attrs(HorizontalLine),
+            "img": self._get_img,
+            "input": self._get_input,
+            # paired tags
+            "ul": self._get_ul,
+            "ol": self._get_ol,
+            "li": self._get_li,
+            "a": self._get_a,
+            "code": self._get_code,
+            "span": self._get_span,
+            "tg-emoji": self._get_tg_emoji,
+            "tg-spoiler": _no_attrs(Spoiler),
+            "pre": self._get_pre,
+            "blockquote": self._get_blockquote,
+            "details": _no_attrs(partial(Blockquote, expandable=True)),
+            "progress": self._get_progress,
+            "meter": self._get_meter,
+            "q": _no_attrs(Quote),
+            "mark": self._get_mark,
+        }
+
+        def _add_map_keys(map, keys, default):
```

Comment:
Please, avoid creating functions inside functions unless you really need closure.